exonum/src/runtime/mod.rs

```diff
@@ -730,3 +727,96 @@ impl<'a> From<(InstanceId, &'a str)> for InstanceDescriptor<'a> {
         InstanceDescriptor { id, name }
     }
 }
+
+/// Execution context extensions required for the Supervisor service implementation.
+#[doc(hidden)]
+#[derive(Debug)]
+pub struct SupervisorExtensions<'a>(ExecutionContext<'a>);
+
+impl<'a> SupervisorExtensions<'a> {
+    /// Marks an artifact as *committed*, i.e., one which service instances can be deployed from.
+    ///
+    /// If / when a block with this instruction is accepted, artifact deployment becomes
+    /// a requirement for all nodes in the network. A node that did not successfully
+    /// deploy the artifact previously blocks until the artifact is deployed successfully.
+    /// If a node cannot deploy the artifact, it panics.
+    pub fn start_artifact_registration(
+        &self,
+        artifact: ArtifactId,
+        spec: Vec<u8>,
+    ) -> Result<(), ExecutionError> {
+        Dispatcher::commit_artifact(self.0.fork, artifact, spec)
+    }
+
+    /// Initiates adding a service instance to the blockchain.
+    ///
+    /// The service is not immediately activated; it activates if / when the block containing
+    /// the activation transaction is committed.    
+    pub fn initiate_adding_service(
+        &mut self,
+        instance_spec: InstanceSpec,
+        constructor: impl BinaryValue,
+    ) -> Result<(), ExecutionError> {
+        self.0
+            .child_context(Some(SUPERVISOR_INSTANCE_ID))
+            .initiate_adding_service(instance_spec, constructor)
+    }
+
+    /// Initiates stopping an active service instance in the blockchain.
+    ///
+    /// The service is not immediately stopped; it stops if / when the block containing
+    /// the stopping transaction is committed.
+    pub fn initiate_stopping_service(&self, instance_id: InstanceId) -> Result<(), ExecutionError> {
+        Dispatcher::initiate_stopping_service(self.0.fork, instance_id)
+    }
+}
+
+/// Collection of unstable execution context features.
+#[doc(hidden)]
+pub trait ExecutionContextUnstable {
+    /// Re-borrows an execution context with the same interface name.
+    fn reborrow(&mut self) -> ExecutionContext<'_>;
```

Comment:
Why is this method unstable?

Reply:
I just move all `pub(crate)` methods to this trait. If you think that I should make some of them public, I will do it.